x/signal/keeper.go

```diff
@@ -78,16 +89,28 @@ func (k Keeper) SignalVersion(ctx context.Context, req *types.MsgSignalVersion)
 	return &types.MsgSignalVersionResponse{}, nil
 }
 
-// TryUpgrade is a method required by the MsgServer interface.
-// It tallies the voting power that has voted on each version.
-// If one version has quorum, it is set as the quorum version
-// which the application can use as signal to upgrade to that version.
+// TryUpgrade is a method required by the MsgServer interface. It tallies the
+// voting power that has voted on each version. If one version has reached a
+// quorum, an upgrade is persisted to the store. The upgrade is used by the
+// application later when it is time to upgrade to that version.
 func (k *Keeper) TryUpgrade(ctx context.Context, _ *types.MsgTryUpgrade) (*types.MsgTryUpgradeResponse, error) {
 	sdkCtx := sdk.UnwrapSDKContext(ctx)
+
+	if k.IsUpgradePending(sdkCtx) {
+		return &types.MsgTryUpgradeResponse{}, types.ErrUpgradePending.Wrapf("can not try upgrade")
+	}
+
 	threshold := k.GetVotingPowerThreshold(sdkCtx)
 	hasQuorum, version := k.TallyVotingPower(sdkCtx, threshold.Int64())
 	if hasQuorum {
-		k.quorumVersion = version
+		if version <= sdkCtx.BlockHeader().Version.App {
+			return &types.MsgTryUpgradeResponse{}, types.ErrInvalidUpgradeVersion.Wrapf("can not upgrade to version %v because it is less than or equal to current version %v", version, sdkCtx.BlockHeader().Version.App)
+		}
+		upgrade := types.Upgrade{
+			AppVersion:    version,
+			UpgradeHeight: sdkCtx.BlockHeader().Height + DefaultUpgradeHeightDelay,
+		}
+		k.setUpgrade(sdkCtx, upgrade)
```

Comment:
[Question] Would it be possible that two validators invoke `TryUpgrade` at two different heights hence resulting in them calculating two different `UpgradeHeight`s? Would it be essential to ensure the entire network end up calculating the same `UpgradeHeight`?

Reply:
Not that I would expect because if validator A and B are on the same chain. Both validators will execute the try upgrade included in the earlier block.